packages/outputarea/src/model.ts

```diff
@@ -484,38 +488,26 @@ namespace Private {
     }
   }
 
-  /**
-   * Remove characters that are overridden by backspace characters.
-   */
-  function fixBackspace(txt: string): string {
-    let tmp = txt;
-    do {
-      txt = tmp;
-      // Cancel out anything-but-newline followed by backspace
-      tmp = txt.replace(/[^\n]\x08/gm, ''); // eslint-disable-line no-control-regex
-    } while (tmp.length < txt.length);
-    return txt;
-  }
-
-  /**
-   * Remove chunks that should be overridden by the effect of
-   * carriage return characters.
-   */
-  function fixCarriageReturn(txt: string): string {
-    txt = txt.replace(/\r+\n/gm, '\n'); // \r followed by \n --> newline
-    while (txt.search(/\r[^$]/g) > -1) {
-      const base = txt.match(/^(.*)\r+/m)![1];
-      let insert = txt.match(/\r+(.*)$/m)![1];
-      insert = insert + base.slice(insert.length, base.length);
-      txt = txt.replace(/\r+.*$/m, '\r').replace(/^.*\r/m, insert);
-    }
-    return txt;
-  }
-
   /*
-   * Remove characters overridden by backspaces and carriage returns
-   */
-  export function removeOverwrittenChars(text: string): string {
-    return fixCarriageReturn(fixBackspace(text));
+   * Concatenate a string to an observable string, handling backspaces.
+   */
+  export function concatenateText(
+    curText: IObservableString,
+    newText: string
+  ): void {
+    for (let index = 0; index < newText.length; index++) {
+      const character = newText[index];
+      if (character === '\b') {
+        // A backspace cannot delete in a previous line.
+        if (
+          curText.text.length > 0 &&
+          curText.text[curText.text.length - 1] != '\n'
+        ) {
+          curText.remove(curText.text.length - 1, curText.text.length);
+        }
+      } else {
+        curText.insert(curText.text.length, character);
+      }
+    }
```

Comment:
This loop will create a text change event for every character in the new text (and accordingly, a refresh of the output in the UI), which is far from optimal.
It is possible to optimize it, but this would probably not be needed if we were directly using Yjs, as we would do all these changes in a single transaction. Unfortunately, ModelDB doesn't seem to support this feature.